packages/flutter/lib/src/widgets/platform_menu_bar.dart

```diff
@@ -0,0 +1,886 @@
+  // PlatformMenuBar to make sure there is only one.
+  BuildContext? _lockedContext;
+
+  @override
+  void clearMenus() => setMenus(<MenuItem>[]);
+
+  @override
+  void setMenus(List<MenuItem> topLevelMenus) {
+    _idMap.clear();
+    final List<Map<String, Object?>> representation = <Map<String, Object?>>[];
+    if (topLevelMenus.isNotEmpty) {
+      int index = 0;
+      for (final MenuItem childItem in topLevelMenus) {
+        if (childItem is! DefaultPlatformMenuDelegateSerializable) {
+          throw UnimplementedError(
+              'Tried to serialize a menu item that was not a DefaultPlatformMenuDelegateSerializer');
+        }
+        representation.addAll((childItem as DefaultPlatformMenuDelegateSerializable)
+            .toChannelRepresentation(this, index: index, count: topLevelMenus.length, getId: _getId));
+        index += 1;
+      }
+    }
+    channel.invokeMethod<void>(_kMenuSetMethod, representation);
+  }
+
+  /// Defines the channel that the [DefaultPlatformMenuDelegate] uses to
+  /// communicate with the platform.
+  ///
+  /// Defaults to [SystemChannels.menu].
+  final MethodChannel channel;
+
+  /// Get the next serialization ID.
+  ///
+  /// This is called by each DefaultPlatformMenuDelegateSerializer when
+  /// serializing a new object so that it has a unique ID.
+  int _getId(MenuItem item) {
+    _serial += 1;
+    _idMap[_serial] = item;
+    return _serial;
+  }
+
+  @override
+  bool debugLockDelegate(BuildContext context) {
+    assert(() {
+      // It's OK to lock if the lock isn't set, but not OK if a different
+      // context is locking it.
+      if (_lockedContext != null && _lockedContext != context) {
+        return false;
+      }
+      _lockedContext = context;
+      return true;
+    }());
+    return true;
+  }
+
+  @override
+  bool debugUnlockDelegate(BuildContext context) {
+    assert(() {
+      // It's OK to unlock if the lock isn't set, but not OK if a different
+      // context is unlocking it.
+      if (_lockedContext != null && _lockedContext != context) {
+        return false;
+      }
+      _lockedContext = null;
+      return true;
+    }());
+    return true;
+  }
+
+  // Handles the method calls from the plugin to forward to selection and
+  // open/close callbacks.
+  Future<void> _methodCallHandler(MethodCall call) async {
+    final int id = call.arguments as int;
+    assert(_idMap.containsKey(id),
+        'Received a menu ${call.method} for a menu item with an ID that was not recognized: $id');
+    if (!_idMap.containsKey(id)) {
+      return;
+    }
+    final MenuItem item = _idMap[id]!;
+    if (item is PlatformMenuItem && call.method == _kMenuSelectedCallbackMethod) {
+      item.onSelected?.call();
+    } else if (item is PlatformMenu && call.method == _kMenuItemOpenedMethod) {
+      item.onOpen?.call();
+    } else if (item is PlatformMenu && call.method == _kMenuItemClosedMethod) {
+      item.onClose?.call();
+    }
```

Comment:
Instead of type checking, could this be implemented polymorphic? E.g. `MenuItem` provides a single interface method (maybe `_performAction`, not sure if it can be private) that is called here. `PlatformMenuItem` and `PlatformMenu` override it to call their actual public. _performAction would have to take some parameters to decide between calling `onOpen` and `onClose`.

Reply:
I still think this type checking is a smell (see also the discussion around the same thing a while ago for the mouse cursor API: https://github.com/flutter/flutter/pull/54171#discussion_r411588713)

Maybe we need a PlatformMenuMixin on MenuItem or something like that and the delegate only deals with PlatformMenuMixin implementors. If you don't need platform support, you can implement MenuItem directly. If you want platform support, well then you have to mix that thing in to get the extra methods it needs to operate on...

Reply:
(same for the Shortcut type checking further down)